pep795.md

```diff
@@ -0,0 +1,1938 @@
+object. (As happens in the next line of the example `f1 = f1 + f2`.)
+
+Let's now bring in functions, modules, and types. To reduce
+clutter, we will not draw the full object graph, since there are
+lots of other default functions e.g. `__eq__` in every type, etc.
+So with some simplifications, the object graph of our example
+above looks like this right before freezing:
+
+```mermaid
+graph TD;
+  A[Stack frame]-- f1 -->B[Fraction instance 1]
+  B-- n -->1
+  B-- d -->3
+  A-- f2 -->D[Fraction instance 2]
+  D-- n -->2
+  D-- d -->7
+  B-- \_\_class__ -->E[Fraction]
+  D-- \_\_class__ -->E
+  1-- \_\_class__ -->int
+  3-- \_\_class__ -->int
+  2-- \_\_class__ -->int
+  7-- \_\_class__ -->int
+  int -- \_\_class__ -->type
+  E-- \_\_init__ -->AA[Function 3]
+  E-- \_\_add__ -->BB[Function 4]
+  E-- \_\_repr__ -->CC[Function 5]
+  E-- \_\_class__ -->type
+  A-- gcd -->DD[Function 1]
+  A-- print_fraction -->EE[Function 2]
+  AA-- gcd -->DD
+  BB-- Fraction -->E
+  int -- superclass -->O[object]
+  E -- superclass -->O
+  O -- \_\_class__ -->type
+```
+
+Note the line from `Function 3` (`Fractions.__init__`) to `Function 1` (`gcd`).
+This is the reference that `__init__` captures to the declaration of
+`gcd` in the enclosing state. If `gcd` was the result of an import,
+the result would have been the same.
+
+To see what should be the effect of freezing, we can go back to
+the first sentence in the second paragraph of this section which
+says that (by default), an object that is immutable may not point
+to an object which is mutable. Thus, we can start by following
+`f1` to its object, colour it "ice blue", and then keep following
+pointers and colour the objects we discover (and lines) to see
+what must be frozen for `freeze(f1)` to succeed. Here is the
+result:
+
+```mermaid
+graph TD;
+  A[Stack frame]-- f1 -->B[Fraction instance 1]
+  B-- n -->1
+  B-- d -->3
+  A-- f2 -->D[Fraction instance 2]
+  D-- n -->2
+  D-- d -->7
+  B-- \_\_class__ -->E[Fraction]
+  D-- \_\_class__ -->E
+  1-- \_\_class__ -->int
+  3-- \_\_class__ -->int
+  2-- \_\_class__ -->int
+  7-- \_\_class__ -->int
+  int -- \_\_class__ -->type
+  E-- \_\_init__ -->AA[Function 3]
+  E-- \_\_add__ -->BB[Function 4]
+  E-- \_\_repr__ -->CC[Function 5]
+  E-- \_\_class__ -->type
+  A-- gcd -->DD[Function 1]
+  A-- print_fraction -->EE[Function 2]
+  AA-- gcd -->DD
+  BB-- Fraction -->E
+  int -- superclass -->O[object]:::imm
+  E -- superclass -->O
+  O -- \_\_class__ -->type
+
+  B:::imm
+  1:::imm
+  3:::imm
+  int:::imm
+  type:::imm
+  E:::imm
+  AA:::imm
+  BB:::imm
+  CC:::imm
+  DD:::imm
+  classDef imm fill:#ADD8E6,stroke:#00f,stroke-width:1px;
+  linkStyle 1,2,6,8,9,12,13,14,15,16,19,20,21,22,23 stroke:#00f,stroke-width:1px,color:blue;
+```
+
+Note that "for `freeze(f1)` to succeed" is not necessarily
+the same as "these objects are made immutable by a call
+to `freeze(f1)`". Later in this document, we will describe
+our proposal for how freezing can propagate.
```

Comment:
This isn't quite clear to me, are you hinting to escape hatches, or that refreezing will succeed even, if the state of the object has not changed?

Reply:
Yeah, I agree. I think this has now been fixed.